gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -857,6 +914,13 @@ protected void afterExecute(Runnable r, Throwable t) {
 
           // Compaction job of a dataset is successful.
           jobRunner.getDataset().setState(COMPACTION_COMPLETE);
+
+          // Add a new MPCompactor job if recompaction is necessary.
+          if (jobRunner.getDataset().needToRecompact()) {
```

Comment:
`MRCompactor.compact()` terminates once all datasets in `this.datasets` are `COMPACTION_COMPLETE`. Since the new dataset is not added to `this.datasets`, `MRCompactor` will terminate and the recompaction job will be killed.

Even if the new dataset is added to `this.datasets`, there will be race conditions: `MRCompactor may determine all datasets are done before the new dataset is added. So the right approach is to reuse the original dataset and never set it to `COMPACTION_COMPLETE` until the recompaction is done.